src/replication/replication_slave_node.cpp

```diff
@@ -47,72 +48,160 @@ namespace cubreplication
+    if (chn.send ((char *) &replication_node::SETUP_REPLICATION_MAGIC, max_len) != css_error_code::NO_ERRORS)
+      {
+	return ER_FAILED;
+      }
 
-    if (m_ctrl_sender != NULL)
+    if (chn.recv ((char *) &expected_magic, max_len) != css_error_code::NO_ERRORS)
       {
-	m_ctrl_sender->stop ();
-	cubthread::get_manager ()->destroy_daemon_without_entry (m_ctrl_sender_daemon);
-	delete m_ctrl_sender;
-	m_ctrl_sender = NULL;
+	return ER_FAILED;
+      }
+
+    if (expected_magic != replication_node::SETUP_REPLICATION_MAGIC)
+      {
+	er_log_debug_replication (ARG_FILE_LINE, "slave_node::setup_protocol error in setup protocol");
+	assert (false);
+	return ER_FAILED;
+      }
+
+    if (chn.recv ((char *) &pos, max_len) != css_error_code::NO_ERRORS)
+      {
+	return ER_FAILED;
       }
+
+    m_source_available_pos = ntohi64 (pos);
+
+    er_log_debug_replication (ARG_FILE_LINE, "slave_node::setup_protocol available pos :%lld",
+			      m_source_available_pos);
+
+    return NO_ERROR;
+  }
+
+  bool slave_node::need_replication_copy (const cubstream::stream_position start_position) const
+  {
+    if (m_source_available_pos >= start_position
+	&& m_source_available_pos - start_position < ACCEPTABLE_POS_DIFF_BEFORE_COPY)
+      {
+	return false;
+      }
+
+    return true;
   }
 
   int slave_node::connect_to_master (const char *master_node_hostname, const int master_node_port_id)
   {
     int error = NO_ERROR;
+
+    if (!m_master_identity.get_hostname ().empty () && m_master_identity.get_hostname () != master_node_hostname)
+      {
+	// master was changed, disconnect from current master and try to connect to new one
+	disconnect_from_master ();
+      }
+
     er_log_debug_replication (ARG_FILE_LINE, "slave_node::connect_to_master host:%s, port: %d\n",
 			      master_node_hostname, master_node_port_id);
 
-    // todo: remove after slave node is able to connect to a different master
     assert (m_transfer_receiver == NULL);
+    assert (m_lc == NULL);
 
     /* connect to replication master node */
     cubcomm::server_channel srv_chn (m_identity.get_hostname ().c_str ());
+    srv_chn.set_channel_name (REPL_ONLINE_CHANNEL_NAME);
 
     m_master_identity.set_hostname (master_node_hostname);
     m_master_identity.set_port (master_node_port_id);
-    error = srv_chn.connect (master_node_hostname, master_node_port_id, SERVER_REQUEST_CONNECT_NEW_SLAVE);
+    error = srv_chn.connect (master_node_hostname, master_node_port_id, COMMAND_SERVER_REQUEST_CONNECT_SLAVE);
     if (error != css_error_code::NO_ERRORS)
       {
 	return error;
       }
 
+    error = setup_protocol (srv_chn);
+    if (error != NO_ERROR)
+      {
+	return error;
+      }
+
+    /* TODO[replication] : last position to be retrieved from recovery module */
+    cubstream::stream_position start_position = log_Gl.m_ack_stream_position;
+
+    if (need_replication_copy (start_position))
+      {
+	/* TODO[replication] : replication copy */
+	assert (false);
```

Comment:
Until it is implemented, we'll have crashes. What do you think of just returning error (and maybe log a message)?